doc/source/user/absolute_beginners.rst

```diff
@@ -0,0 +1,1913 @@
+  >>> print('Number of Columns : ', num_of_columns)
+  Number of Columns :  4
+
+It's also easy to find the total number of elements in your array::
+
+  >>> print(arr.shape[0] * arr.shape[1])
+  12
+
+You can use ``arr.shape()`` with a 1D array as well. If you create this array::
+
+  arr = arr.array([1, 2, 3, 4, 5, 6, 7, 8])
+
+You can print the shape and the length of the array. ::
+
+  >>> print('Shape of 1D array: ', arr.shape)
+  >>> print('Length of 1D array: ', arr.shape[0])
+  Shape of 1D array:  (8,)
+  Length of 1D array:  8
+
+
+You can get the dimensions of an array using ``arr.shape()``. ::
+
+ >>>  # get number of rows in array
+ >>>  num_of_rows2 = arr.shape[0]
+ 
+  >>> # get number of columns in 2D numpy array
+  >>> num_of_columns2 = arr.shape[1]
+ 
+  >>> print('Number of Rows : ', num_of_rows2)
+  >>> print('Number of Columns : ', num_of_columns2)
+  Number of Rows :  3
+  Number of Columns: 4
+
+You can print the total number of elements as well::
+  
+  >>> print('Total number of elements in  array : ', arr.size(arr))
+  Total number of elements in  array :  12
+
+This also works for 3D arrays::
+
+  arr_3d = np.array([ [[1, 1, 1, 1], [2, 2, 2, 2], [3, 3, 3, 3]],
+                      [[4, 4, 4, 4], [5, 5, 5, 5], [6, 6, 6, 6]] ])
+
+You can easily print the size of the axis::
+
+  >>> print('Axis 0 size : ', arr_3d.shape[0]
+  >>> print('Axis 1 size : ', arr_3d.shape[1]
+  >>> print('Axis 2 size : ', arr_3d.shape[2]
+  Axis 0 size :  2
+  Axis 1 size :  3
+  Axis 2 size :  4
+
+You can print the total number of elements:::
+
+  >>> print(arr.size(arr3D))
+  24
+
+You can also use ``arr.size()`` with 1D arrays::
+
+  >>> # Create a 1D array 
+  >>> arr = np.array([1, 2, 3, 4, 5, 6, 7, 8])
+
+  >>> # Determine the length 
+  >>> print('Length of 1D numpy array : ', arr.size)
+
+  Length of 1D numpy array :  8
+
+*Remember that if you check the size of your array and it equals 0, your array
+*is empty.*
+
+Learn more about `finding the size of an array here
+<https://numpy.org/devdocs/reference/generated/numpy.ndarray.size.html>`_ and
+the `shape of an array here
+<https://numpy.org/devdocs/reference/generated/numpy.ndarray.shape.html>`_.
+
+
+Creating matrices 
+-----------------
+
+You can pass Python lists of lists to create a matrix to represent them in
+NumPy. ::
+
+  >>> np.array([[1, 2], [3, 4]])
+
+.. image:: images/np_create_matrix.png
+
+Indexing and slicing operations are useful when you're manipulating matrices::
+
+  data[0, 1]
+  data[1 : 3]
+  data[0 : 2, 0]
+
+.. image:: images/np_matrix_indexing.png
+
+You can aggregate matrices the same way you aggregated vectors::
+
+  data.max()
+  data.min()
+  data.sum()
+
+.. image:: images/np_matrix_aggregation.png
+
+You can aggregate all the values in a matrix and you can aggregate them across
+columns or rows using the ``axis`` parameter::
+  
+  data.max(axis=0)
+  data.max(axis=1)
+
+
+.. image:: images/np_matrix_aggregation_row.png
```

Comment:
![](https://10393-908607-gh.circle-artifacts.com/0/home/circleci/repo/doc/build/html/_images/np_matrix_aggregation_row.png)

Similarly, while this image is great, it is a more accurate visualization for `data.max(axis=0, keepdims=True)`. As it, it seems to suggest that one result is a "row-vector" while the other is a "column-vector" - in fact the shape is indistinguishable.

Perhaps we need some sort of dimensionality marker to indicate which dimension the array is. One option would be some simple square brackets around each row / column, as I do in [this notebook](https://notebooks.azure.com/eric-wieser/projects/numpy-integration):

![image](https://user-images.githubusercontent.com/425260/69885990-d213c100-12d7-11ea-86f6-bb179b0b2fba.png)

Reply:
Very nice! For now, I'm going to defer to @jalammar regarding his images, but I'll be happy to work on whatever is best for everyone.

Reply:
Another option might just be to use gray arrows with numbers next to them, similar to the red arrow above, but on every single axis in this document.

Reply:
@eric-wieser I certainly see your point. This is an oversimplification on my part that I thought long about. I'm not yet happy about a tradeoff that achieves ease of understanding for first-time user, and staying true to how numpy operates. Its most drastic manifestation (and where I feel it's the clunkiest) is in the 3d example:

![numpy-3d](https://user-images.githubusercontent.com/1007956/69897647-ee5b4080-135f-11ea-9ca5-1b2b4e595e92.png)

If you have numpy print out this shape, it would look different from this cube. People come in with the mental model of (x,y,z) mapping to (width, height, depth). But Numpy would print this as (z,y,x) -- and it does that for a very good reason. But that is complexity I wanted to hide from a first-time user. I wanted them to get the intuitive mechanics first, and worry about the idiosyncrasies that numpy is forced to do to be able to operate in an arbitrary number of dimensions later. 

My leaning is to keep the figures as they are, but add a note pointing out this oversimplification. Either the first time we do it, or under each figure where this happens. What do you think?

Reply:
@eric-wieser and I completely agree with you on developing visual language to indicate axes and dimensions. This is one of my attempts (from not-yet-published content I'm developing):

![dimension-indicators](https://user-images.githubusercontent.com/1007956/69897860-c02b3000-1362-11ea-9765-0c74bd814349.png)

Reply:
I've added a note for now!

Reply:
> My leaning is to keep the figures as they are, but add a note pointing out this oversimplification. Either the first time we do it, or under each figure where this happens. What do you think?

This seems quite reasonable. Perhaps as a very clear note (using, e.g., `.. note::` in the reST sources) that there's no such thing as row and column vectors, only 1-D vectors.

I think in the first image in this thread, the `2, 5, 6` result could be rotated for clarity, would be more consistent without taking up much more space. Not a blocker though imho.

> Perhaps we need some sort of dimensionality marker to indicate which dimension the array is. One option would be some simple square brackets around each row / column, as I do in [this notebook](https://notebooks.azure.com/eric-wieser/projects/numpy-integration):

_(there's no "trust this notebook" as commented in the notebook, so I'm just going by the screenshot in this thread)_. This looks quite confusing to be honest. It took me a while to mentally parse, and I'm fairly sure few beginners will be helped by it.

> This is one of my attempts (from not-yet-published content I'm developing):

It's probably hard to add dimensions to an image when there are no extra elements. I can understand the green bar on the side for the extra size-1 dimension. Not sure about the pink bars on the top.

Reply:
> If you have numpy print out this shape, it would look different from this cube. The post is optimized for 2d arrays, so I tried to stay true numpy there (y,x) being (height, width). Going into 3d, people come in with the mental model of (x,y,z) mapping to (width, height, depth). This figure shows (y,x,z) to not break the dimensionality established in the 2d examples. Numpy would print this as (z,y,x) -- and it does that for a very good reason. But that is complexity I wanted to hide from a first-time user. I wanted them to get the intuitive mechanics first, and worry about the idiosyncrasies that numpy is forced to do to be able to operate in an arbitrary number of dimensions later.
> 
> My leaning is to keep the figures as they are, but add a note pointing out this oversimplification. Either the first time we do it, or under each figure where this happens. What do you think?

I agree. I think it's the mental mapping to a cube that matters. Most new users won't pick up on the inconsistency of ordering of `(x, y, z)` probably.

Reply:
> This also shows the dilemma of the previous comment. If this shape is to stay true to numpy, the array on the left would be horizontal and get rotated on the right. The rotation would take the reader's focus away from what we actually want to get across: the new dimension.

But this rotation is really important when it comes to broadcasting. If the reader is not supposed to focus on that, then I'd recommend you use `axis=0` instead.

I'd expect axes of the same color in that image to be broadcastable, but in fact they're not.

> Its most drastic manifestation (and where I feel it's the clunkiest) is in the 3d example:

I don't really see why you can't just swap the text on that image to `ones((2, 4, 3))` - the colors are pretty clear indicators of which elements go where. If you move the `|--2--|` to the top left, then the dimensions continue to read cyclically too.

Reply:
> If the reader is not supposed to focus on that, then I'd recommend you use `axis=0` instead.

that sounds like a nice easy fix.

> I don't really see why you can't just swap the text on that image to `ones((2, 4, 3))`

+1, also nice and easy.

those changes will improve correctness without making the message more complicated.

Note to self/others: this discussion is about images that are not in this PR, so this isn't a blocker.